### PR TITLE
Parsing of cross file upfront, and store in cross-agnostic data structures

### LINKED ISSUE #4332: Separate parsing of cross file from downstream logic
This picks up where #3993 left off, doing for the rest of the file what that did for `[host_machine]`. Firstly, it is good engineering practice to have a clean pipeline by separating parsing from other logic (in code, if not at run time as interleaving can still be achieved via laziness). Secondly, this will do a large portion of the work of abstracting over the cross-vs-native-dependent portions of the code base. Third, this allows easier modifying of the surface syntax, as proposed in e.g. #3972 or #3282 

Here's an incomplete TOOD:

- [x] `[host_machine]` `[target_machine`] `MachineInfo` abstracts over cross file vs environment detection.
  - ~~#3993~~

- [x] `[binaries]` Should abstract over cross file vs environment variable vs default sources of truth. #4326 is convenient for the `PerMachine` abstraction which will allow us to conveniently store tools per machine. Unclear if due to dependence on other platform inference this can happen up front or it needs to happen lazily so as to take place after the former. `_get_compiler` in `environment.py` is a good guide. Unclear whether overlaps with the various lists compilers are stored.
  - ~~#4338~~
  - ~~#4445~~

- [x] `[properties]` Unfortunately defined as open-ended so not much we can do for structure. But we should use `PerMachine` to open the door to differentiating the build host and target machines' additional properties. The compiler options should all be added to per_machine option maps, so everything is taken from `coredata.compiler_options.*` and never `environment.properties` directly. #4626 and #4772 address this last bit.
  - ~~#4445~~
  - ~~#4626~~
  - ~~#4772~~

After all is complete, `self.cross_info` should no longer exist to make ad-hoc inspection impossible. This wasn't attempted in https://github.com/mesonbuild/meson/pull/4010 but in conjunction with that will finish the job of making cross not special. I actually now think it is better to do this before most of that, hence why I am writing this issue now.

CC @nirbheek

### PR DESCRIPTION
**Pull request by Ericson2314**
Continuing what #4338 started. ~~So ignore this until that is merged.~~

Simply put, this gets rid of `environment.cross_info`. Information is instead stored in `environment.binaries` (the `BinaryTable` class has instances now), `environment.machines` as was already done, and `environment.properties`. All of these are `PerMachine` containers, paving the way for a rebased #4010 completely merging cross and native code paths.

In particular the binary lookup code has begun to be consolidated with this, using the same function to assist with looking up in either the cross file or environment variables. There's a lot more consolidation to be done however, that this enables. More net-negative commits on the way!

N.B: `self.environment.is_cross_build() and self.environment.cross_info.need_cross_compiler()` becomes just `self.environment.is_cross_build()`.There's no point distinguishing target =?= build like this, because meson doesn't do anything with the target platform in it's own logic. It's purely there for user logic.

Finally, now that https://github.com/mesonbuild/meson/pull/4216 is merged, this can fold in that code too, and gets rid of `environment.config_info`.

Very close to closing #4332

### CODE REVIEW

File: mesonbuild/dependencies/base.py [hunk 1]
```diff
@@ -1077,13 +1073,15 @@ def description(self):
...
     @staticmethod
-    def from_cross_info(cross_info, name):
-        if name not in cross_info.config['binaries']:
+    def from_cross_file(bt: BinaryTable, name):
+        """Actually from a \"binary table\", which may or may not be cross
+        related. But need to decide how to combine cross and non-cross code
+        path, since it will be a visible change, so leaving this just for the
+        cross change for now.
+        """
+        command = bt.lookup_entry(self, name)
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__

[F821] undefined name 'self'

File: mesonbuild/environment.py [hunk 2]
```diff
@@ -1273,3 +1249,92 @@ def miss_defaulting(self):
...
+        return self.build == self[machine]
+
+class BinaryTable:
+    def __init__(self, binaries = None):
+        self.binaries = binaries
+        if self.binaries is not None:
+            for (key, binary) in self.binaries.items():
+                if not isinstance(binary, (list, str)):
+                    # TODO generalize message
+                    raise MesonException(
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__

[F821] undefined name 'MesonException'

File: mesonbuild/environment.py [hunk 3]
```diff
@@ -1273,3 +1249,92 @@ def miss_defaulting(self):
...
+                        'Invalid type {!r} for binary {!r} in cross file'
+                        ''.format(command, name))
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__

[F821] undefined name 'command'

**Comment by Ericson2314** [comment]
:crossed_fingers: this passes now. Also I set up pkgconfig to be cashed in cross and native alike.

**Comment by Ericson2314** [comment]
CC @dcbaker you would probably be interested in this since it refactors code from #4216.

File: mesonbuild/dependencies/base.py [hunk 4]
```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
...
             else:
-                potential_pkgbin = ExternalProgram.from_bin_list(
-                    environment.cross_info.config['binaries'], 'pkgconfig')
-                if potential_pkgbin.found():
-                    self.pkgbin = potential_pkgbin
+                mlog.debug('Pkg-config binary for %s specified from config file as %s.' % (for_machine, potential_pkgpath))
+                potential_pkgbin = ExternalProgram.from_entry('pkgconfig', potential_pkgpath)
+                if not potential_pkgbin.found():
+                    mlog.debug('Pkg-config %s for machine %s specified at %s but not found.'
+                        % (potential_pkgbin.name, potential_pkgbin.command, for_machine))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E128] continuation line under-indented for visual indent

File: mesonbuild/dependencies/base.py [hunk 5]
```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
...
+                    return potential_pkgbin
+            # Fallback on hard-coded defaults.
+            if environment.machines.matches_build_machine(for_machine):
+                for potential_pkgpath in environment.default_pkgconfig:
+                    potential_pkgbin = self.check_pkgconfig(potential_pkgpath)
+                    if potential_pkgbin is not None:
+                        return potential_pkgbin
+                    else:
+                        mlog.debug('default Pkg-config fallback %s for machine %s specified at %s but not found.'
+                            % (potential_pkgbin.name, for_machine, potential_pkgbin.command))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E128] continuation line under-indented for visual indent

File: mesonbuild/backend/backends.py [hunk 6]
```diff
@@ -318,8 +318,7 @@ def serialize_executable(self, tname, exe, cmd_args, workdir, env={},
                 exe_is_native = True
             is_cross_built = (not exe_is_native) and \
                 self.environment.is_cross_build() and \
-                self.environment.cross_info.need_cross_compiler() and \
```
Comment (Ericson2314) [inline mesonbuild/backend/backends.py]: As per OP `is_cross_build` now implies removed `need_cross_compiler`

File: mesonbuild/dependencies/base.py [hunk 7]
```diff
@@ -390,24 +391,21 @@ def find_config(self, versions=None):
         if not isinstance(versions, list) and versions is not None:
             versions = listify(versions)
 
-        if self.env.is_cross_build() and not self.native:
-            cross_file = self.env.cross_info.config['binaries']
-            try:
-                tools = [cross_file[self.tool_name]]
-            except KeyError:
+        for_machine = MachineChoice.BUILD if self.native else MachineChoice.HOST
+        tool = self.env.binaries[for_machine].lookup_entry(self.tool_name)
```
Comment (Ericson2314) [inline mesonbuild/dependencies/base.py]: This is always a list

File: mesonbuild/dependencies/base.py [hunk 8]
```diff
@@ -390,24 +391,21 @@ def find_config(self, versions=None):
...
+            if self.env.is_cross_build() and not self.native:
                 mlog.warning('No entry for {0} specified in your cross file. '
                              'Falling back to searching PATH. This may find a '
                              'native version of {0}!'.format(self.tool_name))
-                tools = self.tools
-        elif self.tool_name in self.env.config_info.binaries:
-            tools = [self.env.config_info.binaries[self.tool_name]]
-        else:
-            tools = self.tools
+            tools = map(lambda tool: [tool], self.tools)
```
Comment (Ericson2314) [inline mesonbuild/dependencies/base.py]: This is made a list to match the full generality of cross file `[binaries]` entries

File: mesonbuild/dependencies/base.py [hunk 9]
```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
         # stored in the pickled coredata and recovered.
         self.pkgbin = None
```
Comment (Ericson2314) [inline mesonbuild/dependencies/base.py]: The pkg-config code below is now the best version of binary look up I'd next abstract and use everywhere. In particular, it now supports caching in cross and native alike, checks multiple defaults like compilers, and has many detailed debug log messages.

File: mesonbuild/dependencies/ui.py [hunk 10]
```diff
@@ -324,23 +324,10 @@ def _pkgconfig_detect(self, mods, kwargs):
...
-        if self.env.is_cross_build():
-            if 'qmake' in self.env.cross_info.config['binaries']:
-                return ExternalProgram.from_bin_list(self.env.cross_info.config['binaries'], 'qmake')
-        elif self.env.config_info:
-            # Prefer suffixed to unsuffixed version
-            p = ExternalProgram.from_bin_list(self.env.config_info.binaries, 'qmake-' + self.name)
-            if p.found():
-                return p
-            return ExternalProgram.from_bin_list(self.env.config_info.binaries, 'qmake')
-        return ExternalProgram(qmake, silent=True)
```
Comment (Ericson2314) [inline mesonbuild/dependencies/ui.py]: Oh the rest is abstracted away, but I need to add that fallback back

File: mesonbuild/dependencies/base.py [hunk 11]
```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
...
             else:
-                potential_pkgbin = ExternalProgram.from_bin_list(
-                    environment.cross_info.config['binaries'], 'pkgconfig')
-                if potential_pkgbin.found():
-                    self.pkgbin = potential_pkgbin
+                mlog.debug('Pkg-config binary for %s specified from config file as %s.' % (for_machine, potential_pkgpath))
+                potential_pkgbin = ExternalProgram.from_entry('pkgconfig', potential_pkgpath)
+                if not potential_pkgbin.found():
+                    mlog.debug('Pkg-config %s for machine %s specified at %s but not found.'
+                            % (potential_pkgbin.name, potential_pkgbin.command, for_machine))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E128] continuation line under-indented for visual indent

File: mesonbuild/dependencies/base.py [hunk 12]
```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
...
+                    return potential_pkgbin
+            # Fallback on hard-coded defaults.
+            if environment.machines.matches_build_machine(for_machine):
+                for potential_pkgpath in environment.default_pkgconfig:
+                    potential_pkgbin = self.check_pkgconfig(potential_pkgpath)
+                    if potential_pkgbin is not None:
+                        return potential_pkgbin
+                    else:
+                        mlog.debug('default Pkg-config fallback %s for machine %s specified at %s but not found.'
+                                % (potential_pkgbin.name, for_machine, potential_pkgbin.command))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E128] continuation line under-indented for visual indent

File: mesonbuild/dependencies/base.py (hunk 11 above)
Comment (Ericson2314) [reply mesonbuild/dependencies/base.py]: huh, 4 and 8 is incorrect?

File: mesonbuild/environment.py [hunk 13]
```diff
@@ -1071,26 +1077,21 @@ def get_compiler_system_dirs(self):
...
     def get_exe_wrapper(self):
-        if not self.cross_info.need_exe_wrapper():
+        if not self.need_exe_wrapper():
             from .dependencies import EmptyExternalProgram
             return EmptyExternalProgram()
         return self.exe_wrapper
 
-
-class CrossBuildInfo:
+class ConfigFile(dict):
```
Comment (dcbaker) [inline mesonbuild/environment.py]: no, no, no, no, no, no. Do *not* sublcass builtin types. There are so many pitfalls involved, everything may seem right to you, but it will lead to no end of bugs. Just don't do it. Please continue to use composition, keep the self.config variable.

File: mesonbuild/environment.py [hunk 14]
```diff
@@ -1109,61 +1110,38 @@ def parse_datafile(self, filename):
...
                 else:
                     raise EnvironmentException('Malformed value in cross file variable %s.' % entry)
 
-    def has_host(self):
-        return 'host_machine' in self.config
+    @classmethod
+    def _ok_type(cls, i):
+        return isinstance(i, (str, int, bool))
+
+class Properties(dict):
```
Comment (dcbaker) [inline mesonbuild/environment.py]: again, don't subclass builtins

File: mesonbuild/environment.py [hunk 15]
```diff
@@ -1109,61 +1110,38 @@ def parse_datafile(self, filename):
...
-                    self.config[s][entry] = res
+                    self[s][entry] = res
                 else:
                     raise EnvironmentException('Malformed value in cross file variable %s.' % entry)
 
-    def has_host(self):
-        return 'host_machine' in self.config
+    @classmethod
+    def _ok_type(cls, i):
+        return isinstance(i, (str, int, bool))
```
Comment (dcbaker) [inline mesonbuild/environment.py]: at this point this is used in one function, you could either just make it local, or at the very least make it a `@staticmethod` instead of a `@classmethod`

File: mesonbuild/environment.py [hunk 16]
```diff
@@ -1285,7 +1263,26 @@ def libdir_layout_is_win(self):
...
+        and build systems are the same. We don't pass any compilers to
+        detect_cpu_family() here because we always want to know the OS
+        architecture, not what the compiler environment tells us.
+        """
+        if self.system != detect_system():
+            return False
+        true_build_cpu_family = detect_cpu_family({})
+        return \
+            (self.cpu_family == true_build_cpu_family) or \
+            ((true_build_cpu_family == 'x86_64') and (self.cpu_family == 'x86'))
```
Comment (dcbaker) [inline mesonbuild/environment.py]: since you're already using parens here, please use more parens instead of \
```python
return ((self.cpu_family == true_build_cpu_family) or
        ((true_build_cpu_family == 'x86_64') and (self.cpu_family == 'x86')))
```

File: mesonbuild/dependencies/base.py [hunk 17]
```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
...
 
-        # When finding dependencies for cross-compiling, we don't care about
-        # the 'native' pkg-config
-        if self.want_cross:
-            if 'pkgconfig' not in environment.cross_info.config['binaries']:
-                if self.required:
-                    raise DependencyException('Pkg-config binary missing from cross file')
+        for_machine = MachineChoice.HOST \
+            if self.want_cross and environment.is_cross_build() \
+            else MachineChoice.BUILD
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/dependencies/base.py]: I would add that, but in my next even bigger PR, https://github.com/mesonbuild/meson/pull/4010, I get rid of `is_cross` almost everywhere.

File: run_unittests.py [hunk 18]
```diff
@@ -50,11 +50,12 @@
...
-from run_tests import get_builddir_target_args, get_backend_commands, Backend
-from run_tests import ensure_backend_detects_changes, run_configure_inprocess
-from run_tests import run_mtest_inprocess
-from run_tests import FakeBuild, FakeCompilerOptions
+from run_tests import (
+    Backend, FakeBuild, FakeCompilerOptions, ensure_backend_detects_changes,
+    exe_suffix, get_backend_commands, get_builddir_target_args, get_fake_env,
+    get_fake_options, get_meson_script, run_configure_inprocess,
+    run_mtest_inprocess
+)
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply run_unittests.py]: Done, once I push.

File: mesonbuild/dependencies/base.py (hunk 7 above)
Comment (Ericson2314) [reply mesonbuild/dependencies/base.py]: Nope. It is a list of strings.

File: mesonbuild/dependencies/base.py [hunk 19]
```diff
@@ -1078,21 +1093,23 @@ def description(self):
         '''Human friendly description of the command'''
         return ' '.join(self.command)
 
-    @staticmethod
-    def from_bin_list(bins, name):
-        if name not in bins:
+    @classmethod
+    def from_bin_list(cls, bt: BinaryTable, name):
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/dependencies/base.py]: Mmm from my quick checks they seem to do nothing (other than be name-resolved)  in the min supported version of python. Can you measure this? I quite like them for documentation now, and look forward to python 3.7 and beyond making such refactors as I've been doing much easier.

File: mesonbuild/environment.py [hunk 20]
```diff
@@ -331,32 +333,46 @@ def __init__(self, source_dir, build_dir, options):
...
-        self.machines.default_missing()
 
-        if self.coredata.config_files:
-            self.config_info = coredata.ConfigData(
+        # Similar to coredata.compilers and build.compilers, but lower level in
+        # that there is no meta data, only names/paths.
+        self.binaries = PerMachineDefaultable()
+
+        # Misc other properties about each machine.
+        self.properties = PerMachine(Properties(), Properties(), Properties())
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: `PerMachine` is used with many different sorts of things, not just `Properties`, so it's not so easy.

File: mesonbuild/environment.py [hunk 21]
```diff
@@ -331,32 +333,46 @@ def __init__(self, source_dir, build_dir, options):
...
+
+        if self.coredata.config_files is not None:
+            config_info = coredata.ConfigData(
                 coredata.load_configs(self.coredata.config_files))
         else:
-            self.config_info = coredata.ConfigData()
+            config_info = coredata.ConfigData()
+        self.binaries.build = BinaryTable(config_info.binaries)
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: I'd do that in a follow up commit. I could add it in this PR but I figured it's big enough!

File: mesonbuild/environment.py [hunk 22]
```diff
@@ -331,32 +333,46 @@ def __init__(self, source_dir, build_dir, options):
...
+                self.machines.host = MachineInfo.from_literal(config['host_machine'])
+            if 'target_machine' in config:
+                self.machines.target = MachineInfo.from_literal(config['target_machine'])
+
+        self.machines.default_missing()
+        self.binaries.default_missing()
+
+        if self.binaries.host.binaries is not None and \
+           'exe_wrapper' in self.binaries.host.binaries:
+            from .dependencies import ExternalProgram
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: This was how it was done before, see the diff. I bet it is indeed done to avoid a cycle, yuck. I agree it should be fixed, but I rather fix after this; consolidating the way binaries are found will help with moving that elsewhere.

File: mesonbuild/environment.py [hunk 23]
```diff
@@ -879,30 +897,22 @@ def detect_rust_compiler(self, want_cross):
...
-        elif shutil.which("dmd"):
-            exelist = ['dmd']
         else:
-            raise EnvironmentException('Could not find any supported D compiler.')
+            for d in self.default_d:
+                if shutil.which(d):
+                    exelist = [d]
+                    break
+            else:
+                raise EnvironmentException('Could not find any supported D compiler.')
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: `self.binaries.default_missing()` will copy the binaries from build to host in the native case. This and the same things for `self.machines` ensures we always need to check just one of build or host (corresponding to an *intent* of build-time or run-time) thereafter.

File: mesonbuild/environment.py (hunk 15 above)
Comment (Ericson2314) [reply mesonbuild/environment.py]: inlined in followup commit.

File: mesonbuild/dependencies/base.py [hunk 24]
```diff
@@ -498,31 +496,56 @@ def __init__(self, name, environment, kwargs, language=None):
...
-        else:
-            self.pkgbin = PkgConfigDependency.class_pkgbin
+                    return potential_pkgbin
+            # Fallback on hard-coded defaults.
+            if environment.machines.matches_build_machine(for_machine):
+                for potential_pkgpath in environment.default_pkgconfig:
+                    potential_pkgbin = self.check_pkgconfig(potential_pkgpath)
+                    if potential_pkgbin is None:
+                        mlog.debug(
+                             'default Pkg-config fallback %s for machine %s specified at %s but not found.',
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E126] continuation line over-indented for hanging indent

File: mesonbuild/environment.py [hunk 25]
```diff
@@ -1285,7 +1256,25 @@ def libdir_layout_is_win(self):
...
+
+        Can almost always run 32-bit binaries on 64-bit natively if the host
+        and build systems are the same. We don't pass any compilers to
+        detect_cpu_family() here because we always want to know the OS
+        architecture, not what the compiler environment tells us.
+        """
+        if self.system != detect_system():
+            return False
+        true_build_cpu_family = detect_cpu_family({})
+        return ((self.cpu_family == true_build_cpu_family) or
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__ <sub>posted by Sider</sub>

[W504] line break after binary operator

Comment (Ericson2314) [reply mesonbuild/environment.py]: Guess I better go back to backslashes?

File: mesonbuild/environment.py (hunk 13 above)
Comment (Ericson2314) [reply mesonbuild/environment.py]: How about I subclass `SafeConfigParser`?

**Comment by Ericson2314** [comment]
What's up with the swift test?

**Comment by Ericson2314** [comment]
Removed WIP because everything works but suspicious test and some unaddressed review comments.

**Comment by Ericson2314** [comment]
Now depends on #4580, #4581, broken out to debug this and make it slightly smaller.

File: mesonbuild/dependencies/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: I can't find it now, but the reason pep 563 was written (the pep that adds `__future__.annotations` in 3.7) is because at module import time all type annotations have to be resolved, ie, instances of class types are instantiated. It is non-free, it can have a very real cost at startup time.

**Comment by dcbaker** [comment]
Please let me know when you're ready for me to review again.

File: mesonbuild/dependencies/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/dependencies/base.py]: I dove into https://www.python.org/dev/peps/pep-0563/ a bit. The cost they refer to is just in the computation of the annotations; the annotations are never checked. Every annotation i write is just a class. The class exists and must be evaluated, so this should be basically free. It's only stuff like:
```python
_: MyGeneric[MyOtherGeneric[..], MyOtherGeneric[..]]
```
that has anything meaningful to evaluate. And even then it can't be too expensive. It's basically an O(n) with some sublinear scope lookups.

I strongly urge not worrying about this unless we have benchmarks to prove its a problem.

**Comment by Ericson2314** [comment]
@dcbaker what do you think about my responses to your points I didn't change? In particular, what do you think about the tension between `SafeConfigParser` for native files and the manual stuff for cross files?

File: mesonbuild/environment.py (hunk 14 above)
Comment (Ericson2314) [reply mesonbuild/environment.py]: Done!

File: mesonbuild/environment.py (hunk 13 above)
Comment (Ericson2314) [reply mesonbuild/environment.py]: Ah I realized that the old version also used `SafeConfigParser`. They both now have the same transformations.

**Comment by Ericson2314** [comment]
@dcbaker OK this should be ready for another round of review. (Even if CI breaks after these latest changes, I don't think it will be hard to fix.)

**Comment by Ericson2314** [comment]
@dcbaker one question for a follow-up PR is that if the native file also provides `[properties]` like the cross file, that would overlap with the `coredata`-based `get_external_args` and `get_external_link_args`.

I suspect the solution is to differentiate between -for-host and -for-build options on the coredata side too. There's no reason all project-side extra args should be discarded in the cross case. Discarding project flags for custom ones should be an extra choice in native and cross builds alike.

File: mesonbuild/environment.py [hunk 26]
```diff
@@ -1092,99 +1090,88 @@ def get_compiler_system_dirs(self):
...
-        for s in config.sections():
-            self.config[s] = {}
-            for entry in config[s]:
-                value = config[s][entry]
+        for s in parser.sections():
+            section = {}
+            for entry in parser[s]:
+                value = parser[s][entry]
+                # Windows paths...
+                value = value.replace('\\', '\\\\')
```
Comment (Ericson2314) [inline mesonbuild/environment.py]: I copied that from your parser code, but perhaps we should move it so that it only applies to patches. We couldn't want to mess up escape sequences or whatever.

(But if so, let's do that in a separate PR. The tests are fine either way and this matches the current native file semantics.)

**Comment by Ericson2314** [comment]
@dcbaker ping :).

**Comment by dcbaker** [comment]
> I suspect the solution is to differentiate between -for-host and -for-build options on the coredata side too. There's no reason all project-side extra args should be discarded in the cross case. Discarding project flags for custom ones should be an extra choice in native and cross builds alike.

I think so as well. This likely means additional command line changes, as current I believe that -D$lang_args make this assumption. I'm not sure what the right approach there is.

File: mesonbuild/dependencies/base.py (hunk 7 above)
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: So, right blow this we check `if tool is None`, that means that `lookup_entry` returns a scalar type, right?

File: mesonbuild/dependencies/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: I can tell you from experience type annotations before `__future__.annotations` are not free. I bumped a project I had to 3.7 and added the future import and my startup time dropped from several seconds to instantaneous. Apart from that these are useless unless a) we adopt them project wide and b) we start using a tool like mypy to ensure that they're actually right. It also feels sneaky to start using them in a completely unrelated PR.

Please remove them from this PR. Send a separate PR and add them to the whole project if you want to propose using them.

File: mesonbuild/dependencies/ui.py (hunk 10 above)
Comment (dcbaker) [reply mesonbuild/dependencies/ui.py]: have you added the fallback?

File: mesonbuild/environment.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/environment.py]: but the three fields will always be a `Properties()` instance, right?

File: mesonbuild/environment.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/environment.py]: that's fine then, we can fix it later.

File: mesonbuild/environment.py (hunk 26 above)
Comment (dcbaker) [reply mesonbuild/environment.py]: We talked about this when I added that. the conclusion was, don't put escape sequences in your native file entires.

File: mesonbuild/environment.py [hunk 27]
```diff
@@ -1386,3 +1409,26 @@ def parse_entry(cls, entry):
...
+        found.
+
+        First tries looking in explicit map, then tries environment variable.
+        """
+        # Try explict map, don't fall back on env var
+        command = self.binaries.get(name)
+        if command is not None:
+            command = mesonlib.stringlistify(command)
+            # Relies on there being no "" env var
+            evar = self.__class__.evarMap.get(name, "")
```
Comment (dcbaker) [inline mesonbuild/environment.py]: there's no need to use `self.__class__` here, just call `self.evarMap`, pythons MRO lookup will do the right thing.

Comment (dcbaker) [reply mesonbuild/environment.py]: And the other cases in this function as well.

**Comment by dcbaker** [comment]
I still have a few requests that haven't been handled or discussed, but I feel like this is getting pretty close and I like the changes overall, @jpakkane, @nirbheek, opinions?

**Comment by Ericson2314** [comment]
> I think so as well. This likely means additional command line changes, as current I believe that -D$lang_args make this assumption. I'm not sure what the right approach there is.

Awesome. https://github.com/mesonbuild/meson/pull/4626 did the opposite, but that was failing for pragmatic reason too. I started rewriting it last night to instead of the multiple `compiler_options` approach.

File: mesonbuild/dependencies/base.py (hunk 7 above)
Comment (Ericson2314) [reply mesonbuild/dependencies/base.py]: Because this calls `stringlistify` for entries, and `shlex.split` or environment variables, `lookup_entry` always returns either `None` or `List[String]`.

`tools` (note the "s") is always a `List[List[String]]`. The outer list is choices, the inner list is command + arguments.

File: mesonbuild/environment.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: for `self.properties`, but for I use `PerMachine` for many different sort of things. The only requirement is build, host, and target all be the same sort of thing.

[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: In other words, a `PerMachine` type should a parameter. This would be `PerMachine[Properties]`, for example.

**Comment by Ericson2314** [comment]
Oooo likes like when I rebased a new test caught an issue. Should be fixed now.

File: mesonbuild/dependencies/base.py (hunk 7 above)
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: ah, okay. that makes sense.

File: mesonbuild/environment.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/environment.py]: If the fields are always `Properties` instance, then why is it a problem to allow `PerMachine()` to initialize any of the empty fields for you?

**Comment by dcbaker** [comment]
@jpakkane I'd really like your feedback on whether you like the approach or not. I think it's nice and that it's going to make the native and cross file extensions we've been talking about easier to do. If you like the approach (even if you want some implementations changes) I'll base my work on this series.

@Ericson2314 I still have a few comments that are unresolved.

File: mesonbuild/environment.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Ericson2314) [reply mesonbuild/environment.py]: Huh? I'm saying they're *not* always a `Properties` instance. We have `PerMachine[Binaries]` and `PerMachine[MachineInfo]` too. And in my WIP downstream PRs, `PerMachine[Compiler]`, `PerMachine[Dict[UserOption]]`, etc.

File: mesonbuild/environment.py (hunk 16 above)
Comment (Ericson2314) [reply mesonbuild/environment.py]: I had done this, but flake8 was complaining about other things. Sorry I wish I had written it the exact message but it seemed like the non-`\` was actually going to be quite a lot of hoops, so I changed it bak.

**Comment by Ericson2314** [comment]
@dcbaker I think the only remaining issue is @jpakkane cutting the annotation baby? :)

File: mesonbuild/dependencies/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jpakkane) [reply mesonbuild/dependencies/base.py]: AFAICR we do have annotations somewhere in the code base but my grep-fu is failing me. The discussion at the time was very similar to this one, but we chose to allow them since they do work from 3.5 onwards, which is the oldest supported version. Eventually we want to have annotations, though, and adding them incrementally is fine.

That being said if there is a noticeable performance impact then we should be careful. Meson runs fairly slow already.

**Comment by Ericson2314** [comment]
:) and yeah at the first sign of trouble, it's an easy `sed` I'll be happy to do to go remove those annotations.

**Comment by jpakkane** [comment]
I will get this reviewed by tomorrow at the latest.

File: mesonbuild/backend/ninjabackend.py [hunk 28]
```diff
@@ -1381,6 +1381,8 @@ def generate_dynamic_link_rules(self, outfile):
         ctypes = [(self.build.compilers, False)]
         if self.environment.is_cross_build():
             ctypes.append((self.build.cross_compilers, True))
+        else:
+            ctypes.append((self.build.cross_compilers, True))
```
Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py]: The bodies of both the true and false branches are identical. This seems suspicious....

Comment (jpakkane) [reply mesonbuild/backend/ninjabackend.py]: Seems like it.

Comment (Ericson2314) [reply mesonbuild/backend/ninjabackend.py]: Unless you rather me fix the existing issue now, https://github.com/mesonbuild/meson/pull/4010 (which would be the second PR after this one) already rewrites all the code to deduplicate further and rule out this sort of thing for good.
